dm/dm/master/scheduler/scheduler.go

```diff
@@ -1039,6 +1034,73 @@ func (s *Scheduler) RemoveSubTasks(task string, sources ...string) error {
 	return nil
 }
 
+// UpdateSubTasks update the information of one or more subtasks for one task.
+func (s *Scheduler) UpdateSubTasks(ctx context.Context, cfgs ...config.SubTaskConfig) error {
+	if !s.started.Load() {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+	if len(cfgs) == 0 {
+		return nil // no subtasks need to add, this should not happen.
+	}
+	taskNamesM := make(map[string]struct{}, 1)
+	for _, cfg := range cfgs {
+		taskNamesM[cfg.Name] = struct{}{}
+	}
+	taskNames := strMapToSlice(taskNamesM)
+	if len(taskNames) > 1 {
+		// only subtasks from one task supported now.
+		return terror.ErrSchedulerMultiTask.Generate(taskNames)
+	}
+	// check whether exists.
+	for _, cfg := range cfgs {
+		v, ok := s.subTaskCfgs.Load(cfg.Name)
+		if !ok {
+			return terror.ErrSchedulerTaskNotExist.Generate(cfg.Name)
+		}
+		cfgM := v.(map[string]config.SubTaskConfig)
+		_, ok = cfgM[cfg.SourceID]
+		if !ok {
+			return terror.ErrSchedulerSubTaskNotExist.Generate(cfg.Name, cfg.SourceID)
+		}
+	}
+	// check whether in running stage
+	for _, cfg := range cfgs {
+		stage := s.GetExpectSubTaskStage(cfg.Name, cfg.SourceID)
+		if stage.Expect == pb.Stage_Running {
+			return terror.ErrSchedulerSubTaskCfgUpdate.Generate(cfg.Name, cfg.SourceID)
+		}
+	}
+
+	// check by workers todo batch
+	for _, cfg := range cfgs {
+		worker := s.GetWorkerBySource(cfg.SourceID)
+		if worker == nil {
+			return terror.ErrWorkerNoStart
+		}
+		resp, err := worker.checkSubtasksCanUpdate(ctx, &cfg)
+		if err != nil {
+			return err
+		}
+		if !resp.CheckSubtasksCanUpdate.Success {
+			return terror.ErrSchedulerSubTaskCfgUpdate.Generatef("can not update because %s", resp.CheckSubtasksCanUpdate.Msg)
+		}
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
confused about the usage of this lock. we lock the whole interface sometimes, such as `AddSubTasks`, lock nothing sometimes such as `RemoveSubTasks`(maybe there're potential data race, never happens since it's a infrequent call), and partly locked as this interface.

what if the condition is changed after a successfully check but before lock? although should rarely happen since it's a infrequent call.

maybe just lock the whole interface anytime, infrequent call anyway.